geometric_controller/include/geometric_controller/common.h

```diff
@@ -121,4 +132,171 @@ Eigen::Vector4d rot2Quaternion(const Eigen::Matrix3d &R) {
+
+    A(3,3) = dqdot_dq(0,0);
+    A(3,4) = dqdot_dq(0,1);
+    A(3,5) = dqdot_dq(0,2);
+    A(3,6) = dqdot_dq(0,3);
+
+    A(4,3) = dqdot_dq(1,0);
+    A(4,4) = dqdot_dq(1,1);
+    A(4,5) = dqdot_dq(1,2);
+    A(4,6) = dqdot_dq(1,3);
+
+    A(5,3) = dqdot_dq(2,0);
+    A(5,4) = dqdot_dq(2,1);
+    A(5,5) = dqdot_dq(2,2);
+    A(5,6) = dqdot_dq(2,3);
+
+    A(6,3) = dqdot_dq(3,0);
+    A(6,4) = dqdot_dq(3,1);
+    A(6,5) = dqdot_dq(3,2);
+    A(6,6) = dqdot_dq(3,3);
+
+
+    //Velocity
+    dvdot_dq << q.y(),  q.z(),  q.w(), q.x(),
+              -q.x(), -q.w(),  q.z(), q.y(),
+               q.w(), -q.x(), -q.y(), q.z();
+
+    dvdot_dq = 2*norm_thrust*dvdot_dq*q_partial_correction;
+
+    A(7,3) = dvdot_dq(0,0);
+    A(7,4) = dvdot_dq(0,1);
+    A(7,5) = dvdot_dq(0,2);
+    A(7,6) = dvdot_dq(0,3);
+
+    A(8,3) = dvdot_dq(1,0);
+    A(8,4) = dvdot_dq(1,1);
+    A(8,5) = dvdot_dq(1,2);
+    A(8,6) = dvdot_dq(1,3);
+
+    A(9,3) = dvdot_dq(2,0);
+    A(9,4) = dvdot_dq(2,1);
+    A(9,5) = dvdot_dq(2,2);
+    A(9,6) = dvdot_dq(2,3);
+
+    return A;
+}
+
+control_gain_matrix_t B_quadrotor(const state_vector_t& x, const control_vector_t& u)
+{
+   double wx = u(0);
+   double wy = u(1);
+   double wz = u(2);
+   double norm_thrust  = u(3);
+   Eigen::Quaternion<double> q(x(3),x(4),x(5),x(6));
+   Eigen::Matrix<double,3,1> dvdot_dc;
+   Eigen::Matrix<double,4,3> dqdot_dw;
+
+   control_gain_matrix_t B;
+   B.setZero();
+
+   dvdot_dc << 2*(q.w()*q.y() + q.x()*q.z()),
+               2*(q.y()*q.z() - q.w()*q.x()),
+               pow(q.w(),2) - pow(q.x(),2) - pow(q.y(),2) + pow(q.z(),2);
+
+   B(7,3) = dvdot_dc(0);
+   B(8,3) = dvdot_dc(1);
+   B(9,3) = dvdot_dc(2);
+
+   dqdot_dw << -q.x(), -q.y(), -q.z(),
+                q.w(), -q.z(),  q.y(),
+                q.z(),  q.w(), -q.x(),
+               -q.y(),  q.x(),  q.w();
+
+   dqdot_dw = 0.5*dqdot_dw;
+
+   B(3,0) = dqdot_dw(0,0);
+   B(3,1) = dqdot_dw(0,1);
+   B(3,2) = dqdot_dw(0,2);
+
+   B(4,0) = dqdot_dw(1,0);
+   B(4,1) = dqdot_dw(1,1);
+   B(4,2) = dqdot_dw(1,2);
+
+   B(5,0) = dqdot_dw(2,0);
+   B(5,1) = dqdot_dw(2,1);
+   B(5,2) = dqdot_dw(2,2);
+
+   B(6,0) = dqdot_dw(3,0);
+   B(6,1) = dqdot_dw(3,1);
+   B(6,2) = dqdot_dw(3,2);
+
+   return B;
+}
+
+/* Itereation method for discrete model */
+bool solveRiccatiIterationC(const Eigen::MatrixXd &A, const Eigen::MatrixXd &B,
+                            const Eigen::MatrixXd &Q, const Eigen::MatrixXd &R,
+                            Eigen::MatrixXd &P, const double dt = 0.001,
+                            const double &tolerance = 1.E-5,
+                            const uint iter_max = 100000);
+
+bool solveRiccatiIterationC(const Eigen::MatrixXd &A, const Eigen::MatrixXd &B,
+                            const Eigen::MatrixXd &Q, const Eigen::MatrixXd &R,
+                            Eigen::MatrixXd &P, const double dt,
+                            const double &tolerance,
+                            const uint iter_max) {
```

Comment:
Why is this being calculated inline?

Reply:
This a function to calculate the ricatti equations, thats why its a header only function. I can put in the geometric controller class.